From 7fb3712cb1eba18d801fa3fe9d1444d929a83510 Mon Sep 17 00:00:00 2001
From: Camelia Groza <camelia.groza@nxp.com>
Date: Thu, 12 Dec 2019 16:55:02 +0200
Subject: [PATCH 703/741] sdk_dpaa: ls1043a errata: fix potential skb memory
 leak

commit 8b72aacd7d127ab7142db5b8f297b77044984367 from
http://source.codeaurora.org/external/qoriq/qoriq-components/linux

Explicitly remove the old skb outside the w/a, for both successful and
unsuccessful realignments. Make sure the old skb's memory isn't leaked.

Signed-off-by: Camelia Groza <camelia.groza@nxp.com>
Signed-off-by: Zhantao Tang <zhantao.tang@windriver.com>
---
 drivers/net/ethernet/freescale/sdk_dpaa/dpaa_eth_sg.c | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/drivers/net/ethernet/freescale/sdk_dpaa/dpaa_eth_sg.c b/drivers/net/ethernet/freescale/sdk_dpaa/dpaa_eth_sg.c
index d9e9ea5f3f0f..d2ae1380e199 100644
--- a/drivers/net/ethernet/freescale/sdk_dpaa/dpaa_eth_sg.c
+++ b/drivers/net/ethernet/freescale/sdk_dpaa/dpaa_eth_sg.c
@@ -901,7 +901,6 @@ static struct sk_buff *a010022_realign_skb(struct sk_buff *skb,
 	skb_set_network_header(nskb, net_offset);
 	skb_set_transport_header(nskb, trans_offset);
 
-	dev_kfree_skb(skb);
 	return nskb;
 
 err:
@@ -1098,6 +1097,7 @@ int __hot dpa_tx_extended(struct sk_buff *skb, struct net_device *net_dev,
 	int err = 0;
 	bool nonlinear, skb_changed, skb_need_wa;
 	int *countptr, offset = 0;
+	struct sk_buff *nskb;
 
 	/* Flags to help optimize the A010022 errata restriction checks.
 	 *
@@ -1183,7 +1183,7 @@ int __hot dpa_tx_extended(struct sk_buff *skb, struct net_device *net_dev,
 
 		/* Code borrowed from skb_unshare(). */
 		if (skb_cloned(skb) && !skb_need_wa) {
-			struct sk_buff *nskb = skb_copy(skb, GFP_ATOMIC);
+			nskb = skb_copy(skb, GFP_ATOMIC);
 			kfree_skb(skb);
 			skb = nskb;
 			skb_changed = true;
@@ -1218,9 +1218,11 @@ int __hot dpa_tx_extended(struct sk_buff *skb, struct net_device *net_dev,
 			skb_need_wa = true;
 
 		if (unlikely(fm_has_errata_a010022()) && skb_need_wa) {
-			skb = a010022_realign_skb(skb, priv);
-			if (!skb)
+			nskb = a010022_realign_skb(skb, priv);
+			if (!nskb)
 				goto skb_to_fd_failed;
+			dev_kfree_skb(skb);
+			skb = nskb;
 		}
 #endif
 
-- 
2.17.1

